Plugin No Longer Functions

Upon downloading the plugin, the overlay does not appear and the correct item to take for both the grapple and agility portion of the course does not work.

Just encountered the same issue. The plugin would not show up in the plugin list until I restarted runelite after installing it from the hub. Before that it failed with this log error:

```
2025-08-17 12:58:10 BST [pool-3-thread-1] WARN  n.r.c.e.ExternalPluginManager - Unable to start or load external plugin "dorgeshkaan-agility-course"
java.lang.reflect.InvocationTargetException: null
	at java.desktop/java.awt.EventQueue.invokeAndWait(Unknown Source)
	at java.desktop/java.awt.EventQueue.invokeAndWait(Unknown Source)
	at java.desktop/javax.swing.SwingUtilities.invokeAndWait(Unknown Source)
	at net.runelite.client.externalplugins.ExternalPluginManager.refreshPlugins(ExternalPluginManager.java:333)
	at net.runelite.client.util.RunnableExceptionLogger.run(RunnableExceptionLogger.java:41)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.RuntimeException: net.runelite.client.plugins.PluginInstantiationException: java.lang.IllegalStateException: must be called on client thread
	at net.runelite.client.externalplugins.ExternalPluginManager.lambda$refreshPlugins$3(ExternalPluginManager.java:344)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(Unknown Source)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.desktop/java.awt.EventQueue$4.run(Unknown Source)
	at java.desktop/java.awt.EventQueue$4.run(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.desktop/java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.run(Unknown Source)
Caused by: net.runelite.client.plugins.PluginInstantiationException: java.lang.IllegalStateException: must be called on client thread
	at net.runelite.client.plugins.PluginManager.startPlugin(PluginManager.java:459)
	at net.runelite.client.externalplugins.ExternalPluginManager.lambda$refreshPlugins$3(ExternalPluginManager.java:339)
	... 13 common frames omitted
Caused by: java.lang.IllegalStateException: must be called on client thread
	at client.getItemContainer(client.java:33861)
	at com.anderzenn.dorgeshkaanagilitycourse.DorgeshKaanAgilityCourse.hasSpanner(DorgeshKaanAgilityCourse.java:166)
	at com.anderzenn.dorgeshkaanagilitycourse.DorgeshKaanOverlay.updateOverlay(DorgeshKaanOverlay.java:60)
	at com.anderzenn.dorgeshkaanagilitycourse.DorgeshKaanAgilityCourse.updateOverlay(DorgeshKaanAgilityCourse.java:157)
	at com.anderzenn.dorgeshkaanagilitycourse.DorgeshKaanAgilityCourse.startUp(DorgeshKaanAgilityCourse.java:73)
	at net.runelite.client.plugins.PluginManager.startPlugin(PluginManager.java:428)
	... 14 common frames omitted```

--- a/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/DorgeshKaanAgilityCourse.java
+++ b/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/DorgeshKaanAgilityCourse.java
@@ -70,8 +70,7 @@ public class DorgeshKaanAgilityCourse extends Plugin
 		requestedItem1 = config.requestedItem1();
 		requestedItem2 = config.requestedItem2();
 
-		// Plugin startup is not guaranteed to run on the RuneLite client thread.
-		// Any reads of client state (inventory/widgets/player) must be done on the client thread.
+		// Should fix https://github.com/Anderzenn/dorgesh-kaan-agility-course/issues/1
 		clientThread.invokeLater(this::updateOverlay);
 	}
 

--- a/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/DorgeshKaanAgilityCourse.java
+++ b/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/DorgeshKaanAgilityCourse.java
@@ -70,7 +70,9 @@ public class DorgeshKaanAgilityCourse extends Plugin
 		requestedItem1 = config.requestedItem1();
 		requestedItem2 = config.requestedItem2();
 
-		updateOverlay();
+		// Plugin startup is not guaranteed to run on the RuneLite client thread.
+		// Any reads of client state (inventory/widgets/player) must be done on the client thread.
+		clientThread.invokeLater(this::updateOverlay);
 	}
 
 	@Override

--- a/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/DorgeshKaanAgilityCourse.java
+++ b/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/DorgeshKaanAgilityCourse.java
@@ -50,6 +50,9 @@ public class DorgeshKaanAgilityCourse extends Plugin
 	@Inject
 	private SpannerWarningOverlay spannerWarningOverlay;
 
+	@Inject
+	private TurgallHighlightOverlay turgallHighlightOverlay;
+
 	private static final int REGION_ID = 10833;
 
 	private static final Pattern REQUEST_PATTERN = Pattern.compile("The engineer asks you to get a (\\w+) or a (\\w+)");
@@ -64,6 +67,7 @@ public class DorgeshKaanAgilityCourse extends Plugin
 	{
 		overlayManager.add(overlay);
 		overlayManager.add(spannerWarningOverlay);
+		overlayManager.add(turgallHighlightOverlay);
 		//initializeItemPaths();
 
 		// Load saved items in config
@@ -78,6 +82,7 @@ public class DorgeshKaanAgilityCourse extends Plugin
 	protected void shutDown() throws Exception {
 		overlayManager.remove(overlay);
 		overlayManager.remove(spannerWarningOverlay);
+		overlayManager.remove(turgallHighlightOverlay);
 		overlay.clearOverlay();
 		spannerWarningOverlay.clearWarning();
 	}

--- a/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/DorgeshKaanAgilityCourseConfig.java
+++ b/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/DorgeshKaanAgilityCourseConfig.java
@@ -10,6 +10,26 @@ import java.awt.*;
 public interface DorgeshKaanAgilityCourseConfig extends Config
 {
 	@ConfigItem(
+			keyName = "highlightTurgall",
+			name = "Highlight Turgall",
+			description = "Highlight Turgall (NPC ID 2295)"
+	)
+	default boolean highlightTurgall()
+	{
+		return false;
+	}
+
+	@ConfigItem(
+			keyName = "turgallHighlightColour",
+			name = "Turgall highlight colour",
+			description = "Colour used to highlight Turgall"
+	)
+	default Color turgallHighlightColour()
+	{
+		return Color.CYAN;
+	}
+
+	@ConfigItem(
 			keyName = "highlightOption",
 			name = "Highlight Option",
 			description = "Highlight the correct item to pick from the boilers"

--- a/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/TurgallHighlightOverlay.java
+++ b/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/TurgallHighlightOverlay.java
@@ -2,13 +2,17 @@ package com.anderzenn.dorgeshkaanagilitycourse;
 
 import net.runelite.api.Client;
 import net.runelite.api.NPC;
+import net.runelite.api.Point;
 import net.runelite.api.coords.WorldPoint;
 import net.runelite.client.ui.overlay.Overlay;
 import net.runelite.client.ui.overlay.OverlayLayer;
 import net.runelite.client.ui.overlay.OverlayPosition;
 import net.runelite.client.ui.overlay.OverlayUtil;
+import net.runelite.client.util.ColorUtil;
 
 import javax.inject.Inject;
+import java.awt.BasicStroke;
+import java.awt.Color;
 import java.awt.Graphics2D;
 import java.awt.Shape;
 
@@ -56,10 +60,20 @@ public class TurgallHighlightOverlay extends Overlay
 				continue;
 			}
 
-			Shape hull = npc.getConvexHull();
-			if (hull != null)
+			Color border = config.turgallHighlightColour();
+			Color fill = ColorUtil.colorWithAlpha(border, 50);
+
+			Shape area = npc.getConvexHull();
+			if (area == null)
 			{
-				OverlayUtil.renderPolygon(graphics, hull, config.turgallHighlightColour());
+				area = npc.getCanvasTilePoly();
+			}
+
+			if (area != null)
+			{
+				Point mouse = client.getMouseCanvasPosition();
+				OverlayUtil.renderHoverableArea(graphics, area, mouse, fill, border, border.darker());
+				OverlayUtil.renderPolygon(graphics, area, border, fill, new BasicStroke(2));
 			}
 		}
 

--- a/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/TurgallHighlightOverlay.java
+++ b/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/TurgallHighlightOverlay.java
@@ -2,17 +2,13 @@ package com.anderzenn.dorgeshkaanagilitycourse;
 
 import net.runelite.api.Client;
 import net.runelite.api.NPC;
-import net.runelite.api.Point;
 import net.runelite.api.coords.WorldPoint;
 import net.runelite.client.ui.overlay.Overlay;
 import net.runelite.client.ui.overlay.OverlayLayer;
 import net.runelite.client.ui.overlay.OverlayPosition;
 import net.runelite.client.ui.overlay.OverlayUtil;
-import net.runelite.client.util.ColorUtil;
 
 import javax.inject.Inject;
-import java.awt.BasicStroke;
-import java.awt.Color;
 import java.awt.Graphics2D;
 import java.awt.Shape;
 
@@ -60,20 +56,10 @@ public class TurgallHighlightOverlay extends Overlay
 				continue;
 			}
 
-			Color border = config.turgallHighlightColour();
-			Color fill = ColorUtil.colorWithAlpha(border, 50);
-
-			Shape area = npc.getConvexHull();
-			if (area == null)
+			Shape hull = npc.getConvexHull();
+			if (hull != null)
 			{
-				area = npc.getCanvasTilePoly();
-			}
-
-			if (area != null)
-			{
-				Point mouse = client.getMouseCanvasPosition();
-				OverlayUtil.renderHoverableArea(graphics, area, mouse, fill, border, border.darker());
-				OverlayUtil.renderPolygon(graphics, area, border, fill, new BasicStroke(2));
+				OverlayUtil.renderPolygon(graphics, hull, config.turgallHighlightColour());
 			}
 		}
 

--- a/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/TurgallHighlightOverlay.java
+++ b/src/main/java/com/anderzenn/dorgeshkaanagilitycourse/TurgallHighlightOverlay.java
@@ -1,0 +1,69 @@
+package com.anderzenn.dorgeshkaanagilitycourse;
+
+import net.runelite.api.Client;
+import net.runelite.api.NPC;
+import net.runelite.api.coords.WorldPoint;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayUtil;
+
+import javax.inject.Inject;
+import java.awt.Graphics2D;
+import java.awt.Shape;
+
+public class TurgallHighlightOverlay extends Overlay
+{
+	private static final int REGION_ID = 10833;
+	private static final int TURGALL_NPC_ID = 2295;
+
+	private final Client client;
+	private final DorgeshKaanAgilityCourseConfig config;
+
+	@Inject
+	private TurgallHighlightOverlay(Client client, DorgeshKaanAgilityCourseConfig config)
+	{
+		this.client = client;
+		this.config = config;
+
+		setPosition(OverlayPosition.DYNAMIC);
+		setLayer(OverlayLayer.ABOVE_SCENE);
+	}
+
+	@Override
+	public java.awt.Dimension render(Graphics2D graphics)
+	{
+		if (!config.highlightTurgall())
+		{
+			return null;
+		}
+
+		if (client.getLocalPlayer() == null)
+		{
+			return null;
+		}
+
+		int regionId = WorldPoint.fromLocal(client, client.getLocalPlayer().getLocalLocation()).getRegionID();
+		if (regionId != REGION_ID)
+		{
+			return null;
+		}
+
+		for (NPC npc : client.getNpcs())
+		{
+			if (npc == null || npc.getId() != TURGALL_NPC_ID)
+			{
+				continue;
+			}
+
+			Shape hull = npc.getConvexHull();
+			if (hull != null)
+			{
+				OverlayUtil.renderPolygon(graphics, hull, config.turgallHighlightColour());
+			}
+		}
+
+		return null;
+	}
+}
+

